design/BufferOperations.md

```diff
@@ -0,0 +1,149 @@
+The results are initialized in the "pending" state and transition at Javascript task boundary to the "available" state when the implementation can determine the GPU is done using the buffer.
+The mapping operations put the buffer in the mapped state, when a buffer is in the mapped state, only the mapping and unmapping operations on it are allowed.
+In particular a mapped buffer cannot be used in a `WebGPUCommandBuffer` given to `WebGPUQueue.submit`.
+The following must be true or a validation error occurs for mapWrite (resp. mapRead):
+
+ - The buffer must have been created with the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+ - `offset + size` must not overflow and be at most the size of the buffer
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+
+The a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+This operation invalidates all the `WebGPUMappedMemory` created from the buffer and puts the buffer in the unmapped state.
+The buffer must be in the mapped state otherwise a validation error occurs when `unmap` is called.
+
+### WebGPUMappedMemory
+
+`WebGPUMappedMemory` is an object representing a mapped region of a buffer that's both pollable and promise-like.
+
+It can be in one of three states: pending, available and invalidated.
+
+The pollable interface is:
+
+```
+partial interface WebGPUMappedMemory {
+    bool isPending();
+    ArrayBuffer getPointer();
+};
+```
+
+`isPending` return true if the object is in the pending state, false otherwise.
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is acts like a promise as in that it is then-able:
+
+```
+partial interface WebGPUMappedMemory {
+    void then(WebGPUMappedMemorySuccessCallback success,
+              optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead.
+
+When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered.
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
```

Comment:
We often get this argument voiced (the need for immediate upload), but I don't think the API needs special handling for this case, as long as the sync behavior is achievable in general.

Reply:
In this proposal, mapping for writing is asynchronous too, so there is no way to have sync behavior without `Buffer.setSubData`. The alternative would be a `ArrayBuffer? Buffer.mapWriteSync(u32 offset, u32 size);` but I'm not sure this is better. WDYT?

Reply:
Mapping (local) mappable memory should be able to be synchronous if the mappable allocations are actually IPC shmem. Does this design run into issues with how we enqueue copies? It seems iffy, but ideal if it can work reliably synchronously.